File: authentication-client-python/RestClientUtils.py
import httplib
import logging


LOG = logging.getLogger(__name__)


class RestClientUtils:

    @staticmethod
    def verify_response_code(response_code):
        RestClientUtils.check_status(response_code)

    @staticmethod
    def check_status(status):
        raise {
            httplib.OK: LOG.debug("Success operation result code."),
            httplib.NOT_FOUND: RestClientUtils.raise_not_found_error(status),
            httplib.BAD_REQUEST: RestClientUtils.raise_base_request_eror(status),
            httplib.CONFLICT: RestClientUtils.raise_conflict_error(status),
            httplib.UNAUTHORIZED: RestClientUtils.raise_unauthorized_error(status),
            httplib.FORBIDDEN: RestClientUtils.raise_forbidden_error(status),
            httplib.METHOD_NOT_ALLOWED: RestClientUtils.raise_method_not_allowed(status),
            httplib.INTERNAL_SERVER_ERROR: RestClientUtils.raise_internal_server_error(status)
        }.get(status, RestClientUtils.raise_not_supported_error(status))

    @staticmethod
    def raise_not_found_error(status):
        return NotFoundError(status, u'Not found HTTP code was received from gateway server.')

    @staticmethod
    def raise_base_request_eror(status):
        return BadRequestError(status, u'Bad request HTTP code was received from gateway server.')

    @staticmethod
    def raise_conflict_error(status):
        return ConflictError(status, u'Conflict HTTP code was received from gateway server.')

    @staticmethod
    def raise_unauthorized_error(status):
        return UnauthorizedError(status, u'Authorization error code was received from server. ')

    @staticmethod
    def raise_forbidden_error(status):
        return ForbiddenError(status, u'Forbidden HTTP code was received from gateway server')

    @staticmethod
    def raise_method_not_allowed(status):
        return MethodNotAllowed(status, u'Method not allowed code was received from gateway server')

    @staticmethod
    def raise_internal_server_error(status):
        return InternalServerError(status, u'Internal server exception during operation process. ')

    @staticmethod
    def raise_not_supported_error(status):
        return NotSupportedError(status, u'Operation is not supported by gateway server')


class BaseHttpError(Exception):
    def __init__(self, code, msg):

        self.__errorCode = code
        self.__errorMsg = msg

    def code(self):
        return self.__errorCode

    def message(self):
        return self.__errorMsg

    def __str__(self):
        return u"Code: %s \nMessage: %s" % (self.__errorCode, self.__errorMsg)

class BadRequestError(BaseHttpError):

    def __init__(self, code, msg):
        super(BadRequestError,self).__init__(code, msg)

class NotFoundError(BaseHttpError):
    def __init__(self, code, msg):
        super(NotFoundError, self).__init__(code, msg)

class ConflictError(BaseHttpError):
    def __init__(self, code, msg):
        super(ConflictError, self).__init__(code, msg)


class UnauthorizedError(BaseHttpError):
    def __init__(self, code, msg):
        super(UnauthorizedError, self).__init__(code, msg)


class ForbiddenError(BaseHttpError):
    def __init__(self, code, msg):
        super(ForbiddenError, self).__init__(code, msg)


class MethodNotAllowed(BaseHttpError):
    def __init__(self, code, msg):
        super(MethodNotAllowed, self).__init__(code, msg)


class InternalServerError(BaseHttpError):
    def __init__(self, code, msg):
        super(InternalServerError, self).__init__(code, msg)


class NotSupportedError(BaseHttpError):
    def __init__(self, code, msg):
        super(NotSupportedError, self).__init__(code, msg)

File: authentication-client/authentication-client-java/src/main/java/co/cask/cdap/security/authentication/client/basic/RestClientUtils.java
/*
 * Copyright 2014 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package co.cask.cdap.security.authentication.client.basic;

import com.google.common.base.Charsets;
import com.google.common.io.CharStreams;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.entity.ContentType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.Charset;
import javax.ws.rs.BadRequestException;
import javax.ws.rs.ForbiddenException;
import javax.ws.rs.InternalServerErrorException;
import javax.ws.rs.NotAllowedException;
import javax.ws.rs.NotAuthorizedException;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.NotSupportedException;

/**
 * The utility class for working with rest clients.
 */
public final class RestClientUtils {
  private static final Logger LOG = LoggerFactory.getLogger(RestClientUtils.class);

  private RestClientUtils() {
  }

  /**
   * Utility method for analysis http response status code and throw appropriate Java API Exception
   *
   * @param response {@link org.apache.http.HttpResponse} http response
   */
  public static void verifyResponseCode(HttpResponse response) {
    int code = response.getStatusLine().getStatusCode();
    switch (code) {
      case HttpStatus.SC_OK:
        LOG.debug("Success operation result code.");
        break;
      case HttpStatus.SC_NOT_FOUND:
        throw new NotFoundException("Not found HTTP code was received from gateway server.");
      case HttpStatus.SC_CONFLICT:
        throw new BadRequestException("Conflict HTTP code was received from gateway server.");
      case HttpStatus.SC_BAD_REQUEST:
        throw new BadRequestException("Bad request HTTP code was received from gateway server.");
      case HttpStatus.SC_UNAUTHORIZED:
        throw new NotAuthorizedException(response);
      case HttpStatus.SC_FORBIDDEN:
        throw new ForbiddenException("Forbidden HTTP code was received from gateway server");
      case HttpStatus.SC_METHOD_NOT_ALLOWED:
        throw new NotAllowedException(response.getStatusLine().getReasonPhrase());
      case HttpStatus.SC_INTERNAL_SERVER_ERROR:
        throw new InternalServerErrorException("Internal server exception during operation process.");
      case HttpStatus.SC_NOT_IMPLEMENTED:
      default:
        throw new NotSupportedException("Operation is not supported by gateway server");
    }
  }
}
